Drop the return and corruptionPossible arguments from CValidationState::DoS, and rename to ::Reject

Empact opened the pull request:
These arguments were always false. With this, DoS and Invalid return false always, as with error.

Empact committed:
Drop return argument from CValidationState::Invalid

It's false in every case.

Empact committed:
Drop return argument from CValidationState::DoS

It was always false. Note error() always returns false.

instagibbs commented:
concept ACK, anything that gets rid of some of that nightmare argument list is a plus

Empact commented:
Note other PRs in this area: #11523 #11639

practicalswift commented:
Nice cleanup!

Concept ACK

laanwj commented:
Concept ACK

conscott reviewed (commented):
Test ACk 829f5461cd9ba850bededff78640ce1807ffcee2

MarcoFalke commented:
Needs rebase due to "witness reserved value" rename.

jnewbery commented:
utACK 829f5461cd9ba850bededff78640ce1807ffcee2

Empact commented:
Rebased & ran clang-format.

jnewbery commented:
utACK 228e3f9965b66f23c611a944eed9454ca5d433eb

MarcoFalke commented:
utACK 228e3f9

sipa commented:
This is a bit scary, due to the large number of optional arguments to `DoS` I'd be more comfortable reviewing if it was clear that all call sites have been touched. You could do this by either writing it as a scripted diff, or by renaming the method `DoS` to something else.

MarcoFalke commented:
Good catch. It seems it must be renamed, since a scripted diff could still lead to silent merge conflicts.

Empact committed:
Drop unused default arguments to CValidationState::DoS and ::Invalid

Empact commented:
I have a somewhat different take in mind, will push that up in a bit.

Empact committed:
Make the level explicit for CValidationState::Invalid

Empact committed:
Use ::SetCorruptionPossible to declare corruption together with ::DoS

This builds the object using two calls when possiblyCorrupt argument is
set. This makes the interface more simple & explicit by making
this literal argument self-explanatory, and by not separating the
message strings.

Empact commented:
* Introduced ::Corrupt to sit alongside ::Invalid
* Dropped DoS in favor of the two above
* Removed another random boolean argument in the process

So the only accidental compilable change would require using ::Invalid, where the method signature has changed by adding a leading required int.

jnewbery commented:
Travis job timed out. I've restarted it.

jnewbery commented on src/consensus/validation.h:
```diff
@@ -37,24 +37,21 @@ class CValidationState {
-             unsigned int chRejectCodeIn=0, const std::string &strRejectReasonIn="",
-             bool corruptionIn=false,
-             const std::string &strDebugMessageIn="") {
+    bool Invalid(int level, unsigned int chRejectCodeIn, const std::string& strRejectReasonIn, const std::string& strDebugMessageIn = "")
```
I think you can make `strDebugMessageIn` a required argument for this and `Corrupt()` below, since all calls to both functions specify a `strDebugMessageIn`. That removes the final optional argument.

Empact replied:
It's interesting to consistently provide validation debug information though, that's a project I'm up for.

jnewbery commented on src/consensus/validation.h:
```diff
@@ -37,24 +37,21 @@ class CValidationState {
         strRejectReason = strRejectReasonIn;
-        corruptionPossible = corruptionIn;
         strDebugMessage = strDebugMessageIn;
         if (mode == MODE_ERROR)
```
Probably scope creep so feel free to ignore, but since you're changing things around so much, it'd be clearer to me to change this to:

```
if (mode != MODE_ERROR)
```

and unify the `return false`s at the end of the function.

MarcoFalke commented:
See also discussion in #11523

Empact commented:
Ok as of 4757699 I've
* replaced the ::Corrupt/::Invalid split with a chaining call to ::SetCorruptionPossible
* dropped ::Invalid in favor of ::DoS (because I wanted to use Invalid for other purposes below)

As of ebc666a I have
* Split the DoS calls along the REJECT_CODES
* Privatized DoS entirely - note this involves changing the reject code of 2 errors from 0 to REJECT_INVALID - this seemed like an oversight, but lmk if this change is problematic.

This PR is looking more and more inspired by #11523 :P

jnewbery commented:
The scope of this PR seems to have crept quite substantially! Can you update the PR title to indicate what the current branch is actually doing.

I think I like some of the new validation state interface that you've added here, but you should get the opinion of @TheBlueMatt , who's rewritten that interface here: #11639.

TheBlueMatt commented:
Indeed, this seems to have grown into something rather duplicative with #11639 with a slightly different approach. I didn't do the removal of the "ret" argument there, but I think it (appropriately) goes further than this by removing the "DoS" field entirely (its really, really dumb that its there). Further, I think the reason list in this PR is incomplete - malleated should be a seaparate reason from Invalid, etc and I believe having a invalid-due-to-fork distinction is probably important. Not sure what the best way to move forward is here - it seems like #11639 generated no review, and now has been rewritten at least once, plus other work in this area.

Empact commented:
Fair enough! The goal was to take an incremental approach and I'd rather stick with that, so I backed out the last two commits described above, and renamed ::DoS to ::Reject, in light of @sipa's comments.

Empact commented:
Moved this back to sharing a common base with 228e3f9, for easier diffing / review.

TheBlueMatt commented:
This seems like a great PR to make use of the scripted-diff infrastructure. re: coflicts with #11639 it might make sense to skip the "level explicit in ::Invalid" and "SetCorruptionPossible" changes (especially the SetCorruptionPossible change - I'm not convinced its the right approach - CorruptionPossible is a separate state which should be indicated as such, not a flag which should be added on top of an existing state).

Empact committed:
scripted-diff: Rename CValidationState::DoS to ::Reject

To protect against silent merge conflicts.

-BEGIN VERIFY SCRIPT-
find ./src -type f \( -name *.cpp -o -name *.h \) -exec sed -i -e 's/\bDoS(/Reject(/' {} \;
-END VERIFY SCRIPT-

Empact commented:
Planning to split this down over a few PRs to simplify review, e.g. making maximum used of scripted-diff.

Empact commented:
Closing in favor of #12976, which attempts to incorporate all the feedback from here.
